### DISCUSSION TITLE
Proposal for Assert.Count

### DISCUSSION THREAD

**Discussion by stefanedwards** [xunit/xunit#2145] [Ideas]
### Proposal:

We already have `Assert.Empty` and `Assert.Single` to test if a collection is either empty or only contains a single element. I am repeatedly writing `Assert.Equals(int, collection.Count())` to test if my collections contain a specific number of elements (but more than one).

When `Assert.Empty` and `Assert.Single` fails, we get informative messages. For collection sizes larger than 1, when testing with `Assert.Equal`, we do not get informative messages. Only that two integers do not match. Finding the corresponding fail test is then only possible by referring to the line number.

The proposal is there fore to add `Assert.Count` or `Assert.Size` to `CollectionAsserts` which upon failure outputs a message similar to,

> The collection was expected to contain {0} element, but {1} elements were found.

where {0} and {1} corresponds to the expected and actual number of elements, respectively.

### Implementation:

Some enumerable collections contain a `Count` or `Length` property, which can be leveraged to avoid enumerating each and every element, see [CA1829](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1829?view=vs-2019), as does Linq's `Count()` command (https://github.com/dotnet/runtime/issues/33216#issuecomment-595255001).

A similar approach can be applied for `Assert.Count`/`Assert.Size`.

For raising exceptions, the corresponding exception class could be `CountExceptions` or `SizeExceptions`.

If accepted, I would gladly contribute with this implementation, given some guidance (especially as the exceptions are stored in a separate repository from the `CollectionAsserts` class).

**Comment by bradwilson** [comment]
I'm not particularly against the idea, but I'd like to understand the usage.

By comparison, I always use `Assert.Collection` which not only counts the items, but gives them to you so you can assert on them. I would rarely consider "does this list have `n` items in it?" to be a sufficient test, and would rather want to look at the items in the list to make sure that it had the _right_ items.

**Reply by stefanedwards** [reply]
I will start with an argument by induction; we have `Assert.Empty` for testing 0 items and `Assert.Single` for testing 1 item without regard to what that object is. So by that sense, testing say 2 items, without regard to what that object is, is not far stretch. And so on, up til testing for `n` items.

My tasks are mainly testing complex data structures, where different test scenarios focuses on different sets of properties. E.g. a single test might test 2 or 3 different properties depending on if it is the 1st, 2nd, or nth element. If I had to use a comparer for this, it might be 1 or 2 comparers per test, in addition to writing the actual expected items. `Assert.Collection` could be used in these cases, but I find it loses readability if the number of items/properties tested is more than an hand full.

An example for using `Assert.Count` would be in this actual example:

```csharp
Assert.Equal(4, outputContainer.Count());
var items = outputContainer.OrderBy(o => o.FirstOut).ToList();
Assert.Null(items[0].PreviousId);
Assert.Null(items[0].LastOut2FirstOut);
Assert.Null(items[0].LastOut2LastOut);
Assert.Null(items[0].PreviousProductNumber);
Assert.Null(items[0].PreviousWorkOrderId);

Assert.NotNull(items[3].PreviousId);
Assert.Equal(2, items[3].LastOut2FirstOut);
Assert.Equal(4, items[3].LastOut2LastOut);
Assert.Equal("1", items[3].PreviousProductNumber);
Assert.Null(items[3].PreviousWorkOrderId);
```

**Reply by bradwilson** [reply]
FWIW, `Assert.Single` returns the single item, because it is assumed that you want to test the values inside the object.

**Reply by bradwilson** [reply]
This is how I'd write this personally:

```csharp
Assert.Collection(
    outputContainer.OrderBy(o => o.FirstOut),
    first =>
    {
        Assert.Null(first.PreviousId);
        Assert.Null(first.LastOut2FirstOut);
        Assert.Null(first.LastOut2LastOut);
        Assert.Null(first.PreviousProductNumber);
        Assert.Null(first.PreviousWorkOrderId);
    },
    second => { },  // Some comment about why I don't care
    third => { },   // Some comment about why I don't care
    fourth =>
    {
        Assert.NotNull(fourth.PreviousId);
        Assert.Equal(2, fourth.LastOut2FirstOut);
        Assert.Equal(4, fourth.LastOut2LastOut);
        Assert.Equal("1", fourth.PreviousProductNumber);
        Assert.Null(fourth.PreviousWorkOrderId);
    }
);
```

I'd be interested to understand how often the size of the list is so large that this is impractical, thus necessitating `Assert.Count` rather than just leaving it as `Assert.Equal(...)`.

**Reply by stefanedwards** [reply]
I appreciate the discussion. I've definitely learned some useful tips about Xunit. 

I do not have any further comments, other than to let this proposal sit and see if there are others who need this feature.

Thanks for your time.

**Reply by bradwilson** [reply]
Yep, happy to leave it open and see what others have to say/vote. 😄

**Reply by stefanedwards** [reply]
Here's an actual case where an `Assert.Count` could improve readability and code consistency. My initial test (thanks for making me aware of `Assert.Collection`!) was testing 3 elements in a collection, where I wanted to test the number of elements in each item, with respectively 2, 1, and 1 elements.

```
Assert.Collection(mapped,
    first => {
        Assert.Equal(2, first.Steps.Where(x => x.TestType == "helium_test").Count());              // <----
        Assert.False(first.Accepted);
    },
    second => {
        Assert.Equal(1, second.Steps.Where(x => x.TestType == "helium_test").Count());        // <----
        Assert.True(second.Accepted);
    },
    third => {
        Assert.Equal(1, third.Steps.Where(x => x.TestType == "helium_test").Count());            // <----
        Assert.True(third.Accepted);
    }
);
```
I was aiming for a bit of consistency, so the 2nd and 3rd case was tested in same manner as 1st.

But oh no, said Xunit, `Do not use Assert.Equal() to check for collection size.`. There's only one fix, unless I wrap it with `#pragma warning disable xUnit2013`:

```
Assert.Collection(mapped,
    first => {
        Assert.Equal(2, first.Steps.Where(x => x.TestType == "helium_test").Count());
        Assert.False(first.Accepted);
    },
    second => {
        Assert.Single(second.Steps.Where(x => x.TestType == "helium_test"));
        Assert.True(second.Accepted);
    },
    third => {
        Assert.Single(third.Steps.Where(x => x.TestType == "helium_test"));
        Assert.True(third.Accepted); 
    }
);
```

To quote Uncle Roger, "Hiyaaaa...."

**Reply by bradwilson** [reply]
You can globally disable the specific analyzer if you disagree with it. 😄 

I'm still not convinced that Assert.Count is more than just an esoteric case, so I'm going to mark this as answered. If people come by and provide enough support for it, I'll consider adding it at a later time.

**Reply by Der-Kraken** [reply]
I found this topic because I wanted the "Assert.Count" feature too. But the coding style of @bradwilson is really great.
The Assert.Collection is a good thing (if you know it).

_By the way:
The Assert.Count feature would be nice for dynamically/generic Asserts (not hard coded ones)._
